iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/pipeconsensus/PipeConsensusAsyncConnector.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.pipe.connector.protocol.pipeconsensus;
+
+import org.apache.iotdb.commons.conf.CommonConfig;
+import org.apache.iotdb.commons.conf.CommonDescriptor;
+import org.apache.iotdb.db.pipe.connector.protocol.thrift.async.IoTDBDataRegionAsyncConnector;
+import org.apache.iotdb.pipe.api.event.Event;
+import org.apache.iotdb.pipe.api.event.dml.insertion.TabletInsertionEvent;
+import org.apache.iotdb.pipe.api.event.dml.insertion.TsFileInsertionEvent;
+import org.apache.iotdb.pipe.api.exception.PipeException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Iterator;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+
+public class PipeConsensusAsyncConnector extends IoTDBDataRegionAsyncConnector {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(PipeConsensusAsyncConnector.class);
+
+  private static final String ENQUEUE_EXCEPTION_MSG =
+      "Timeout: PipeConsensusConnector offers an event into transferBuffer failed, because transferBuffer is full";
+
+  private static final CommonConfig COMMON_CONFIG = CommonDescriptor.getInstance().getConfig();
+
+  private final BlockingQueue<Event> transferBuffer =
+      new LinkedBlockingDeque<>(COMMON_CONFIG.getPipeConsensusEventBufferSize());
+
+  /** Add an event to transferBuffer, whose events will be asynchronizedly transfer to receiver. */
+  private boolean addEvent2Buffer(Event event) {
```

Comment:
Do we need to increase the reference count of the event when adding it to buffer? And also decrease reference count when removing?

Reply:
In current design, we will increase the reference count when transfer method and decrease it in callback method of AsyncThrift RPC. but I think it's better to manage reference count in the way you said, since it's semantically clearer.  I will try to fix it as you suggested~

Reply:
> Do we need to increase the reference count of the event when adding it to buffer? And also decrease reference count when removing?

@DanielWang2035 In current design, we will increase the reference count when transfer method and decrease it in callback method of AsyncThrift RPC. but I think it's better to manage reference count in the way you said, since it's semantically clearer. I will try to fix it as you suggested~